chunk_downloader.go

```diff
@@ -664,3 +701,32 @@ func copyChunkStream(body io.Reader, rows chan<- []*string) error {
 	}
 	return nil
 }
+
+// A ResultBatch object encapsulates a function that retrieves a subset of rows in a result set.
+type ResultBatch struct {
+	rec                []array.Record
+	idx                int
+	scd                *snowflakeChunkDownloader
+	funcDownloadHelper func(context.Context, *snowflakeChunkDownloader, int) error
+}
+
+// Fetch returns an array of records representing a chunk in the query
+func (rb *ResultBatch) Fetch() ([]array.Record, error) {
+	if len(rb.scd.ResultBatches) == 1 {
+		return rb.scd.ResultBatches[0].rec, nil
+	}
+	err := rb.funcDownloadHelper(context.Background(), rb.scd, rb.idx)
+	if err != nil {
+		return nil, err
+	}
+	return rb.scd.ResultBatches[rb.idx].rec, nil
```

Comment:
perhaps fetch should only return `error` since we're putting the `[]array.Record` inside the `.rec` field already

Reply:
rationale: we probably don't want to store the full record both on the `ResultBatch` object AND upon returning it and thus saving it in memory into another object. Unless we want to do both and leave it up to the client to decide - I like the return `error` and exporting `Rec` personally.

Reply:
I exposed `.Rec` in new commit (amended). These msgs are not applicable anymore.